File: go/days/day02/day02.go
package day02

import (
	"bufio"
	"bytes"
	"embed"
	"fmt"
)

var limits = map[string]int{
	"red":   12,
	"green": 13,
	"blue":  14,
}

//go:embed games.txt
var gamesInput embed.FS

func Run() (int, int, error) {
	var part1Result, part2Result int

	data, err := gamesInput.Open("games.txt")
	if err != nil {
		return -1, -1, fmt.Errorf("opening calibration document: %w", err)
	}
	defer data.Close()

	scanner := bufio.NewScanner(data)

	for scanner.Scan() {
		if scanner.Err() != nil {
			return -1, -1, fmt.Errorf("scanning calibration document: %w", err)
		}

		lineBytes := scanner.Bytes()

		part1Result += possibleGames(lineBytes)
		part2Result += maxPowerGames(lineBytes)
	}

	return part1Result, part2Result, nil
}

func possibleGames(line []byte) int {
	gameID, rounds, _ := bytes.Cut(line, []byte(":"))

	results := bytes.Split(rounds, []byte(";"))

	for i := range results {
		cubes := bytes.Split(results[i], []byte(","))

		for j := range cubes {
			trimmed := bytes.TrimSpace(cubes[j])
			amount, color, _ := bytes.Cut(trimmed, []byte(" "))

			amt := byteToInt(amount)

			if amt > limits[string(color)] {
				return 0
			}
		}
	}

	successfulAmt, _ := bytes.CutPrefix(gameID, []byte("Game "))

	return byteToInt(successfulAmt)
}

func maxPowerGames(line []byte) int {
	_, rounds, _ := bytes.Cut(line, []byte(":"))

	results := bytes.Split(rounds, []byte(";"))

	highestCubeCount := make(map[string]int, 3)

	for i := range results {
		cubes := bytes.Split(results[i], []byte(","))

		for j := range cubes {
			trimmed := bytes.TrimSpace(cubes[j])
			amount, color, _ := bytes.Cut(trimmed, []byte(" "))
			colorStr := string(color)

			amt := byteToInt(amount)

			stored, ok := highestCubeCount[colorStr]
			if !ok || amt > stored {
				highestCubeCount[colorStr] = amt
			}
		}
	}

	result := 1
	for _, v := range highestCubeCount {
		result *= v
	}

	return result
}

func byteToInt(input []byte) int {
	var result int

	for i := range input {
		result = result*10 + int(input[i]-'0')
	}

	return result
}

File: zig/days/day02/day02.zig
const std = @import("std");
const part1 = @import("./part1.zig");
const part2 = @import("./part2.zig");
const filepath = "days/day02/games.txt";
const allocator = std.heap.page_allocator;

const day02_result = struct {
    part1: u32,
    part2: u64,
};

pub fn run() anyerror!day02_result {
    const file = try std.fs.cwd().openFile(filepath, .{});
    defer file.close();

    var buf_reader = std.io.bufferedReaderSize(1024, file.reader());
    const reader = buf_reader.reader();

    var buffer = try allocator.alloc(u8, 1024);
    defer allocator.free(buffer);

    var part1_result: u32 = 0;
    var part2_result: u64 = 0;

    while (try reader.readUntilDelimiterOrEof(buffer, '\n')) |line| {
        part1_result += try part1.possibleGames(line);
        part2_result += try part2.maxPowerGames(line);
    }

    const result = day02_result{
        .part1 = part1_result,
        .part2 = part2_result,
    };

    return result;
}
